explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+
+On top of that, the current `autofdo` version in Ubuntu (`0.19-3build3`, at the time of this writing) is not recent enough to process the `perfdata` files we will generate.  There is a PPA with a newer version of `autofdo` package for Ubuntu Noble [here](https://launchpad.net/~sergiodj/+archive/ubuntu/autofdo).  If you are running another version of Ubuntu and want to install a newer version of `autofdo`, you will need to build the software manually (please refer to the [upstream repository](https://github.com/google/autofdo) for further instructions).
+
+## A simple PGO scenario: `openssl speed`
+
+PGO makes sense when your software is CPU-bound, i.e., when it performs CPU intensive work and is not waiting on I/O, for example.  OpenSSL has a built-in benchmark command called `openssl speed`, which tests the performance of its cryptographic algorithms.  This is excellent for PGO because there is practically no I/O involved, and we are just constrained by how fast the CPU can run.
+
+### Running OpenSSL tests
+
+In order to measure the performance impact of PGO, it is important to perform several tests using an OpenSSL binary without PGO and then with it.
+
+The first thing we have to do is to do several runs of `openssl speed` *before* enabling PGO, so that we have a baseline to compare with.  As explained in the sections above, `autofdo` does not handle stripped binaries nor `dwz`-processed debug information, so we need to make
+After confirming that everything looks OK, we are ready to start the benchmark.  Let's run the command:
+
+```bash
+$ openssl speed -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+This will run benchmark tests for `md5`, `sha512`, `rsa2048` and `aes-256-cbc`.  Each test will last 60 seconds, and they will involve calculating as many cryptographic hashes from `N`-byte chunks of data (with `N` varying from 16 to 16384) with each algorithm (with the exception of `rsa2048`, whose performance is measured in signatures/verifications per second).  By the end, you should see a report like the following:
+
+```bash
+The 'numbers' are in 1000s of bytes per second processed.
+type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
+sha512           91713.74k   366632.96k   684349.30k  1060512.22k  1260359.00k  1273277.92k
+aes-256-cbc    1266018.36k  1437870.33k  1442743.14k  1449933.84k  1453336.84k  1453484.99k
+md5             140978.58k   367562.53k   714094.14k   955267.21k  1060969.81k  1068296.87k
+                  sign    verify    sign/s verify/s
+rsa 2048 bits 0.000173s 0.000012s   5776.8  86791.9
+```
+
+### Building OpenSSL for profiling
+
+Before we are able to profile OpenSSL, we need sure that we compile the software in a way that the generated program meets the requirements.  In our case, you can use the `file` command after building the binary to make sure that it has not been stripped:
```

Comment:
we know, but a random person might not immediately know how to "make sure that it has not been stripped:".

maybe you want to refer what you watch out for in the file output e.g. ` with debug_info, not stripped`

Reply:
ACK.  Done.